### DISCUSSION TITLE
How I can recognize if there is no internet connection while loading

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3700] [General]
Hi! I use the following code to download url. But it doesnt throw any exception if internet is disconnected (now I see that just by getting empty source html code of the loaded webpage):

`public virtual void LoadUrlCore(string url)
        {
            (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode = -1;

            if (String.IsNullOrEmpty(url)) return;
          

            var tcs = new TaskCompletionSource<bool>();
            EventHandler<LoadingStateChangedEventArgs> handler = null;
            handler = (sender, args) =>
            {
                
                if (!args.IsLoading)
                {

                    Browser.LoadingStateChanged -= handler;

                    tcs.TrySetResultAsync(true);
                }
            };
        
            Browser.LoadingStateChanged += handler;

            if (!string.IsNullOrEmpty(url))
            {
                Browser.Load(url);
            }
            
            if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
                throw new TimeoutException("Время ожидания загрузки страницы истекло");

            if ((Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != 200 &&
                (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != -1)
                throw new Exception("Неверный код ответа страницы: " + (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode);

        }`

**Comment by amaitland** [comment]
The LoadError event should be called and provide the relevant error code. 

I'd suggest you can use the new [LoadUrlAsync](http://cefsharp.github.io/api/91.1.x/html/M_CefSharp_IWebBrowser_LoadUrlAsync.htm) method, you can get errors and http status code from the response.

You'll need version 91 for this method. 

https://github.com/cefsharp/CefSharp/blob/4e16f3b16dbdedb6f145cbb665b5819906072e21/CefSharp/WebBrowserExtensions.cs#L256

**Reply by assopri** [reply]
Unfortunately I cannot move right now to the 91 version. Any ideas to catch load error and let my method caller know about it?

**Reply by amaitland** [reply]
You can update your code to implement LoadError just as in LoadUrlAsync.

**Reply by assopri** [reply]
Thank you! Done smth like this:
`public virtual void LoadUrlCore(string url)
        {
            (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode = -1;

            if (String.IsNullOrEmpty(url)) return;
          

            var tcs = new TaskCompletionSource<bool>();

            EventHandler<LoadErrorEventArgs> loadErrorHandler = null;
            EventHandler<LoadingStateChangedEventArgs> loadingStateChangeHandler = null;

            // https://github.com/cefsharp/CefSharp/blob/4e16f3b16dbdedb6f145cbb665b5819906072e21/CefSharp/WebBrowserExtensions.cs#L256
            loadErrorHandler = (sender, args) =>
            {
                if (args.ErrorCode == CefErrorCode.Aborted)
                {
                    return;
                }

                Browser.LoadError -= loadErrorHandler;
                Browser.LoadingStateChanged -= loadingStateChangeHandler;

                tcs.TrySetResultAsync(new LoadUrlAsyncResponse(args.ErrorCode, -1));
            };

            loadingStateChangeHandler = (sender, args) =>
            {
                
                if (!args.IsLoading)
                {
                    Browser.LoadError -= loadErrorHandler;
                    Browser.LoadingStateChanged -= loadingStateChangeHandler;

                    tcs.TrySetResultAsync(true);
                }
            };

            Browser.LoadError += loadErrorHandler;
            Browser.LoadingStateChanged += loadingStateChangeHandler;

            if (!string.IsNullOrEmpty(url))
            {
                Browser.Load(url);
            }
            
            if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
                throw new TimeoutException("Время ожидания загрузки страницы истекло");

            if ((Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != 200 &&
                (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != -1)
                throw new Exception("Неверный код ответа страницы: " + (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode);

        }`
However getting such error:
![image](https://user-images.githubusercontent.com/73235464/126610442-f9d92373-b90b-4124-8d55-7dc0ff90b2b6.png)

**Reply by amaitland** [reply]
As you are blocking on the `Task` returned from the `TaskCompletionSource` (I personally don't recommend calling `Task.Wait`, rewrite your code to use `await`) it probably doesn't matter which thread the continuation happens on, so just call `TrySetResult` instead.

**Reply by assopri** [reply]
Thank you! I changed like this:

`private string _lastLoadError = "";

        public virtual void LoadUrlCore(string url)
        {
            _lastLoadError = "";

            (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode = -1;

            if (String.IsNullOrEmpty(url)) return;
          

            var tcs = new TaskCompletionSource<bool>();

            EventHandler<LoadErrorEventArgs> loadErrorHandler = null;
            EventHandler<LoadingStateChangedEventArgs> loadingStateChangeHandler = null;

            // https://github.com/cefsharp/CefSharp/blob/4e16f3b16dbdedb6f145cbb665b5819906072e21/CefSharp/WebBrowserExtensions.cs#L256
            loadErrorHandler = (sender, args) =>
            {
                if (args.ErrorCode == CefErrorCode.Aborted)
                {
                    return;
                }

                Browser.LoadError -= loadErrorHandler;
                Browser.LoadingStateChanged -= loadingStateChangeHandler;
                _lastLoadError = args.ErrorCode.ToString();

                tcs.TrySetResult(true);// new LoadUrlAsyncResponse(args.ErrorCode, -1));
            };

            loadingStateChangeHandler = (sender, args) =>
            {
                
                if (!args.IsLoading)
                {
                    //Browser.LoadError -= loadErrorHandler;
                    Browser.LoadingStateChanged -= loadingStateChangeHandler;

                    tcs.TrySetResultAsync(true);
                }
            };

            Browser.LoadError += loadErrorHandler;
            Browser.LoadingStateChanged += loadingStateChangeHandler;

            if (!string.IsNullOrEmpty(url))
            {
                Browser.Load(url);
            }
            
            if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
                throw new TimeoutException("Время ожидания загрузки страницы истекло");

            if(!String.IsNullOrEmpty(_lastLoadError)) throw new TimeoutException("Ошибка загрузки страницы: "+ _lastLoadError);

            if ((Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != 200 &&
                (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != -1)
                throw new Exception("Неверный код ответа страницы: " + (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode);

        }
`

However I still face situation that LoadingStateChanged is hooked earlier than LoadError.

**Reply by amaitland** [reply]
What exactly is stopping you from upgrading to version 91?

**Reply by assopri** [reply]
Ok I already upgraded. How can I update my original function to use LoadUrlAsync . How I extract LoadError from it's return value without await and changing my caller function ( LoadUrlCore ) to async?

```

public virtual void LoadUrlCore(string url)
        {
            _lastLoadError = "";

            (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode = -1;

            if (String.IsNullOrEmpty(url)) return;
          
            var tcs = new TaskCompletionSource<bool>();

            EventHandler<LoadingStateChangedEventArgs> loadingStateChangeHandler = null;

            loadingStateChangeHandler = (sender, args) =>
            {                
                if (!args.IsLoading)
                {
                    Browser.LoadingStateChanged -= loadingStateChangeHandler;

                    tcs.TrySetResultAsync(true);
                }
            };

            Browser.LoadingStateChanged += loadingStateChangeHandler;

            if (!string.IsNullOrEmpty(url))
            {
                Browser.LoadUrlAsync(url);
            }
            
            if (!tcs.Task.Wait(TimeSpan.FromMilliseconds(LoadTimeoutMs)))
                throw new TimeoutException("Время ожидания загрузки страницы истекло");

            if(!String.IsNullOrEmpty(_lastLoadError)) throw new TimeoutException("Ошибка загрузки страницы: "+ _lastLoadError);

            if ((Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != 200 &&
                (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode != -1)
                throw new Exception("Неверный код ответа страницы: " + (Browser.RequestHandler as CustomRequestHandler).LastResponceStatusCode);

        }
```

**Reply by amaitland** [reply]
You can obtain the Error code and http status code from the response.  

http://cefsharp.github.io/api/91.1.x/html/M_CefSharp_IWebBrowser_LoadUrlAsync.htm
http://cefsharp.github.io/api/91.1.x/html/T_CefSharp_LoadUrlAsyncResponse.htm

**Reply by assopri** [reply]
Thank you! That works just fine
